drivers/crypto/crypto_mchp_sha_g1.c

```diff
@@ -0,0 +1,348 @@
+	{CRYPTO_HASH_ALGO_SHA384, SHA_MR_ALGO_SHA384, 384 / BITS_PER_BYTE, 128},
+	{CRYPTO_HASH_ALGO_SHA512, SHA_MR_ALGO_SHA512, 512 / BITS_PER_BYTE, 128},
+};
+
+static struct crypto_mchp_sha_session mchp_sha_sessions[2];
+static struct k_sem mchp_sha_session_sem;
+
+static inline void mchp_sha_set_input(sha_registers_t *const regs, const unsigned char *data,
+				      int len)
+{
+	const uint32_t *data_u32 = (const uint32_t *)data;
+	uint32_t i = 0, v;
+
+	while (len >= sizeof(uint32_t)) {
+		v = data_u32[i];
+		if (i < ARRAY_SIZE(regs->SHA_IDATAR)) {
+			regs->SHA_IDATAR[i] = v;
+		} else {
+			regs->SHA_IODATAR[i - ARRAY_SIZE(regs->SHA_IDATAR)] = v;
+		}
+		i++;
+		len -= sizeof(uint32_t);
+	}
+
+	if (len > 0) {
+		v = data[i * sizeof(uint32_t) + 0];
+		if (len > 1) {
+			v |= (uint32_t)data[i * sizeof(uint32_t) + 1] << 8;
+		}
+		if (len > 2) {
+			v |= (uint32_t)data[i * sizeof(uint32_t) + 2] << 16;
+		}
+		if (i < ARRAY_SIZE(regs->SHA_IDATAR)) {
+			regs->SHA_IDATAR[i] = v;
+		} else {
+			regs->SHA_IODATAR[i - ARRAY_SIZE(regs->SHA_IDATAR)] = v;
+		}
+	}
+}
+
+static inline void mchp_sha_get_output(sha_registers_t *const regs, unsigned char *data, int len)
+{
+	uint32_t *data_u32 = (uint32_t *)data;
+	uint32_t i;
+
+	for (i = 0; i < (len / sizeof(uint32_t)) && i < ARRAY_SIZE(regs->SHA_IODATAR); i++) {
+		data_u32[i] = regs->SHA_IODATAR[i];
+	}
+}
+
+static int mchp_sha_wait_data_rdy(sha_registers_t *const regs)
+{
+	uint32_t timeout = 1;
+
+	while ((regs->SHA_ISR & SHA_ISR_DATRDY_Msk) == 0) {
+		if (timeout-- == 0) {
+			LOG_ERR("MCHP SHA wait data ready timeout");
+			return -ETIMEDOUT;
+		}
+		k_sleep(K_USEC(5));
+	}
+
+	return 0;
+}
+
+static int mchp_sha_process(sha_registers_t *const regs,
+			    struct crypto_mchp_sha_algo_cfg const *algo_cfg,
+			    const unsigned char *data, unsigned int len, unsigned char *digest)
+{
+	int block_size = algo_cfg->block_size;
+	unsigned int processed = 0;
+	int ret;
+
+	regs->SHA_CR = SHA_CR_SWRST_Msk;
+	regs->SHA_MR = algo_cfg->sha_mr_algo | SHA_MR_SMOD_AUTO_START | SHA_MR_PROCDLY_LONGEST;
+	regs->SHA_CR = SHA_CR_FIRST_Msk;
+	regs->SHA_MSR = len;
+	regs->SHA_BCR = len;
+
+	if (len == 0) {
+		/* For the empty message, automatic padding is not quired in this driver
+		 * (SHA_MSR.MSGSIZE and SHA_BCR.BYTCNT are configured to 0). The block to
+		 * be processed are the padded part (a one bit, then  '1' followed by zeor
+		 * bits) and then the message length.
+		 */
+		regs->SHA_IDATAR[0] = FIRST_WORD_4_PADDING_EMPTY_MSG;
+		for (int i = 1; i < block_size / sizeof(uint32_t); i++) {
+			regs->SHA_IDATAR[i] = 0;
+		}
+		goto out;
+	}
+
+	while (len - processed >= block_size) {
+		/* Write the block to be processed in the Input Data Registers */
+		mchp_sha_set_input(regs, &data[processed], block_size);
+		ret = mchp_sha_wait_data_rdy(regs);
+		if (ret < 0) {
+			return ret;
+		}
+
+		processed += block_size;
+	}
+
+	if (len - processed) {
```

Comment:
you could make it a boolean expression (as recommended in the guideline - I think).